// amp-filter/src/utils/constants.js
var constants = {
    FILTER_ID_ORGS: 'organizations',
    FILTER_NAME_DONOR: 'Donor',
    FILTER_ID_SECTOR: 'sectors',
    FILTER_ID_PROGRAM: 'programs',
    FILTER_ID_LOCATION: 'locations',
    FILTER_ID_DATE: 'date',
    FILTER_ID_COMPUTED_YEAR: 'computed-year',
    LEVEL_ONE: 1,
    LEVEL_TWO: 2,
    LEVEL_THREE: 3,
    LEVEL_FOUR: 4,
    LEVEL_FIVE: 5,
    ALL_FILTERS_URL: '/rest/nifilters',
    ACTIVITY_BUDGET_LIST: 'ActivityBudgetList',
    TYPE_OF_ASSISTANCE: 'type-of-assistance',
    MODE_OF_PAYMENT: 'mode-of-payment',
    EXPENDITURE_CLASS: 'expenditure-class',
    FINANCING_INSTRUMENT: 'financing-instrument',
    FUNDING_STATUS: 'funding-status',
    EFFECTIVE_FUNDING_DATE: 'effective-funding-date',
    FUNDING_CLOSING_DATE: 'funding-closing-date',
    CONCESSIONALITY_LEVEL: 'concessionality-level',
    ON_OFF_TREASURY_BUDGET: 'On/Off/Treasury Budget',
    STATUS: 'Status',
    PERFORMANCE_ALERT_LEVEL: 'performance-alert-level',
    PERFORMANCE_ALERT_TYPE: 'performance-alert-type',
    APPROVAL_STATUS: 'approval-status',
    PROGRAMS: 'Programs',
    SECTORS: 'Sectors',
    DONOR: 'Donor',
    ROLE: 'Role',
    LOCATIONS: 'Locations',
    PLEDGES_SECTORS: 'Pledges Sectors',
    PLEDGES_SECONDARY_SECTORS: 'Pledges Secondary Sectors',
    PLEDGES_PROGRAMS: 'Pledges Programs',
    PLEDGES_LOCATIONS: 'Pledges Locations',
    PLEDGES_DONORS: 'Pledges Donors',
    COMPONENT_TYPE_ALL: 'ALL',
    FIELD_DATA_TYPE_TREE: 'TREE',
    FIELD_DATA_TYPE_TEXT: 'TEXT',
    DATE_RANGE_VALUES: 'DATE-RANGE-VALUES',
    YEAR_SINGLE_VALUE: 'YEAR-SINGLE-VALUE',
    CONTEXT_DASHBOARD: 'DASHBOARD',
    CONTEXT_GIS: 'GIS'
};

module.exports = constants;
