// amp-filter/src/utils/constants.js
var constants = {
    FILTER_ID_ORGS: 'organizations',
    FILTER_NAME_DONOR: 'Donor',
    FILTER_ID_SECTOR: 'sectors',
    FILTER_ID_PROGRAM: 'programs',
    FILTER_ID_LOCATION: 'locations',
    FILTER_ID_DATE: 'date',
    FILTER_ID_COMPUTED_YEAR: 'computed-year',
    LEVEL_ONE: 1,
    LEVEL_TWO: 2,
    LEVEL_THREE: 3,
    LEVEL_FOUR: 4,
    LEVEL_FIVE: 5,
    LEVEL_SIX: 6,
    ALL_FILTERS_URL: '/rest/nifilters',
    ACTIVITY_BUDGET_LIST: 'ActivityBudgetList',
    TYPE_OF_ASSISTANCE: 'type-of-assistance',
    TYPE_OF_ASSISTANCE_NAME: 'Type Of Assistance',
    MODE_OF_PAYMENT: 'mode-of-payment',
    EXPENDITURE_CLASS: 'expenditure-class',
    FINANCING_INSTRUMENT: 'financing-instrument',
    FINANCING_INSTRUMENT_NAME: 'Financing Instrument',
    FUNDING_STATUS: 'funding-status',
    EFFECTIVE_FUNDING_DATE: 'effective-funding-date',
    FUNDING_CLOSING_DATE: 'funding-closing-date',
    CONCESSIONALITY_LEVEL: 'concessionality-level',
    ACTIVITY_BUDGET: 'Activity Budget',
    STATUS: 'Status',
    PERFORMANCE_ALERT_LEVEL: 'performance-alert-level',
    PERFORMANCE_ALERT_TYPE: 'performance-alert-type',
    APPROVAL_STATUS: 'approval-status',
    ACTIVITY_STATUS: 'Activity Status',
    PROGRAMS: 'Programs',
    SECTORS: 'Sectors',
    DONOR: 'Donor',
    ROLE: 'Role',
    LOCATIONS: 'Locations',
    PLEDGES_LOCATIONS: 'Pledges Locations',
    PLEDGES_DONORS: 'Pledges Donors',
    COMPONENT_TYPE_ALL: 'ALL',
    FIELD_DATA_TYPE_TREE: 'TREE',
    FIELD_DATA_TYPE_TEXT: 'TEXT',
    DATE_RANGE_VALUES: 'DATE-RANGE-VALUES',
    YEAR_SINGLE_VALUE: 'YEAR-SINGLE-VALUE',
    CONTEXT_DASHBOARD: 'DASHBOARD',
    CONTEXT_GIS: 'GIS',
    FUNDING_ORGANIZATIONS: 'Funding Organizations',
    ACTIVITY: 'Activity',
    ALL_AGENCIES: 'All Agencies',
    FINANCIALS: 'Financial',
    OTHERS: 'Other',
    LOCATION: 'Location',
    SECTOR: 'Sector',
    GROUP: "group",
    PLEDGES: 'pledges',
    DEFAULT_COMPUTED_YEAR_NAME: 'Current',
    START_DATE_TEMPLATE: '{year}-01-01',
    END_DATE_TEMPLATE: '{year}-12-31',
    YEAR_PLACEHOLDER: '{year}',
    UNASSIGNED: 'Unassigned',
    ADM_LEVEL_0: 'administrative-level-0',
    ADM_LEVEL_1: 'administrative-level-1',
    ADM_LEVEL_2: 'administrative-level-2',
    ADM_LEVEL_3: 'administrative-level-3',
    ADM_LEVEL_4: 'administrative-level-4',
    PLEDGES_ADM_LEVEL_0: 'pledges-administrative-level-0',
    PLEDGES_ADM_LEVEL_1: 'pledges-administrative-level-1',
    PLEDGES_ADM_LEVEL_2: 'pledges-administrative-level-2',
    PLEDGES_ADM_LEVEL_3: 'pledges-administrative-level-3',
    PLEDGES_ADM_LEVEL_4: 'pledges-administrative-level-4'
};

module.exports = constants;
