### Eclipse Workspace Patch 1.0
#P ampp
Index: sites/all/modules/contrib/advanced_mail_reroute/advanced_mail_reroute.module
===================================================================
RCS file: /wwwroot/repository/drupal/ampp/sites/all/modules/contrib/advanced_mail_reroute/advanced_mail_reroute.module,v
retrieving revision 1.4
diff -u -r1.4 advanced_mail_reroute.module
--- sites/all/modules/contrib/advanced_mail_reroute/advanced_mail_reroute.module	17 Jan 2012 07:31:29 -0000	1.4
+++ sites/all/modules/contrib/advanced_mail_reroute/advanced_mail_reroute.module	17 Jan 2012 07:32:08 -0000
@@ -87,7 +87,7 @@
  * Implementation of hook_mail_alter()
  */
 function advanced_mail_reroute_mail_alter(&$message) {  
-  $num_rows = db_result(db_query("SELECT COUNT(*) FROM {advanced_mail_reroute_rules} WHERE mailkey='%s'", $message['id']));
+  $num_rows = db_query("SELECT COUNT(*) FROM {advanced_mail_reroute_rules} WHERE mailkey=':mailkey'", array(':mailkey' => $message['id']))->fetchField();
   
   // fetch or create new rule
   if ($num_rows > 0) {
@@ -100,7 +100,9 @@
       'email' => ''
     );
         
-    db_query("INSERT INTO {advanced_mail_reroute_rules} (mailkey, reroute_rule, email) VALUES ('%s', %d, '')", $message['id'], $rule['reroute_rule']);
+    db_insert("advanced_mail_reroute_rules")
+    ->fields(array('mailkey', 'reroute_rule', 'email'))
+    ->values(array($message['id'], $rule['reroute_rule']));
   }
 
   // apply override rule
@@ -452,25 +454,24 @@
 //                           Private Functions
 //=============================================================================
 function _advanced_mail_reroute_log($mailkey, $from, $to, $reroute_rule, $reroute_email = '') {
-    db_query("INSERT INTO {advanced_mail_reroute_log} (timestamp, mailkey, from_email, to_email, reroute_rule, reroute_email)
-      VALUES (%d, '%s', '%s', '%s', %d, '%s')", time(), $mailkey, $from, $to, $reroute_rule, $reroute_email);
+    db_insert("advanced_mail_reroute_log")
+    ->fields(array('timestamp', 'mailkey', 'from_email', 'to_email', 'reroute_rule', 'reroute_email'))
+    ->values(array(time(), $mailkey, $from, $to, $reroute_rule, $reroute_email));
 }
 
-function _advanced_mail_reroute_reroute(&$message, $reroute_to) {
-  $message['body'] = is_array($message['body']) ? implode("\n\n", $message['body']) : $message['body'];
-  
+function _advanced_mail_reroute_reroute(&$message, $reroute_to) {  
   global $base_url;
   $message['subject'] = 'Rerouted email: '. $message['subject'];
   
   // Format copied from reroute_email module: http://drupal.org/project/reroute_email
-  $body_header  = "This email was rerouted.\n";
+  $body_header = "This email was rerouted.\n";
   $body_header .= "Web site: @site\n";
   $body_header .= "Mail key: @key\n";
   $body_header .= "Originally to: <@to>\n";
   $body_header .= "-----------------------\n";
 
-  $body_header = t($body_header, array('@site' => $base_url, '@to' => $message['to'], '@key' => $message['id']));
+  $body[] = t($body_header, array('@site' => $base_url, '@to' => $message['to'], '@key' => $message['id']));
 
-  $message['body'] = $body_header . (is_array($message['body']) ? implode("\n\n", $message['body']) : $message['body']);
+  $message['body'] = array_merge($body, $message['body']);
   $message['to'] = $reroute_to;
-}
+}
\ No newline at end of file
