### Eclipse Workspace Patch 1.0
#P ampp
Index: sites/all/modules/contrib/captcha/captcha.module
===================================================================
RCS file: /wwwroot/repository/drupal/ampp/sites/all/modules/contrib/captcha/captcha.module,v
retrieving revision 1.1
diff -u -r1.1 captcha.module
--- sites/all/modules/contrib/captcha/captcha.module    16 Jan 2012 09:02:11 -0000  1.1
+++ sites/all/modules/contrib/captcha/captcha.module    20 Jan 2012 08:35:19 -0000
@@ -531,7 +531,7 @@
           array(':csid' => $posted_captcha_sid)
         )->fetchField();
         if ($expected_captcha_token !== $posted_captcha_token) {
-          drupal_set_message(t('CAPTCHA session reuse attack detected.'), 'error');
+          form_set_error('captcha_response', t('CAPTCHA session reuse attack detected.'));
           // Invalidate the CAPTCHA session.
           $posted_captcha_sid = NULL;
         }
