Index: sites/all/modules/contrib/link/link.module
===================================================================
RCS file: /wwwroot/repository/drupal/ampp/sites/all/modules/contrib/link/link.module,v
retrieving revision 1.1
diff -u -r1.1 link.module
--- sites/all/modules/contrib/link/link.module  19 Jan 2012 16:53:56 -0000  1.1
+++ sites/all/modules/contrib/link/link.module  2 Feb 2012 19:02:06 -0000
@@ -248,6 +248,7 @@
  * Implements hook_field_validate().
  */
 function link_field_validate($entity_type, $entity, $field, $instance, $langcode, $items, &$errors) {
+  //print_r($items); exit;
   $optional_field_found = FALSE;
   if ($instance['settings']['validate_url'] !== 0 || is_null($instance['settings']['validate_url']) || !isset($instance['settings']['validate_url'])) {
     foreach ($items as $delta => $value) {
@@ -357,19 +358,20 @@
       && $item['url'] === $instance['default_value'][$delta]['url']
       && !$instance['required'])) {
     // Validate the link.
-    if (link_validate_url(trim($item['url'])) == FALSE) {
-      form_set_error($field['field_name'] .']['. $delta .'][url', t('Not a valid URL.'));
-    }
+     
+    if (link_validate_url(trim($item['url'])) == FALSE) { //echo $field['field_name'] . ' - ' . $delta; exit;
+      form_set_error($field['field_name'] .'][und]['. $delta .'][url', t('Not a valid URL.'));
+   }
     // Require a title for the link if necessary.
     if ($instance['settings']['title'] == 'required' && strlen(trim($item['title'])) == 0) {
-      form_set_error($field['field_name'] .']['. $delta .'][title', t('Titles are required for all links.'));
+      form_set_error($field['field_name'] .'][und]['. $delta .'][title', t('Titles are required for all links.'));
     }
   }
   // Require a link if we have a title.
   if ($instance['settings']['url'] !== 'optional'
       && strlen(isset($item['title']) ? $item['title'] : NULL) > 0
       && strlen(trim($item['url'])) == 0) {
-    form_set_error($field['field_name'] .']['. $delta .'][url', t('You cannot enter a title without a link url.'));
+    form_set_error($field['field_name'] .'][und]['. $delta .'][url', t('You cannot enter a title without a link url.'));
   }
   // In a totally bizzaro case, where URLs and titles are optional but the field is required, ensure there is at least one link.
   if ($instance['settings']['url'] === 'optional'
