// amp-boilerplate/index.js
var fs = require('fs');
var _ = require('underscore');
var $ = require('jquery');
var Backbone = require('backbone');

var bootstrap_enabled = (typeof $().modal == 'function');

if (bootstrap_enabled) {
  require('bootstrap/dist/js/bootstrap');
}

var MenuView = require('./src/menu-view.js');

/* example of use
 * this.menus = new Menus({
 *   translator: this.translator,
 *   caller: 'GIS'
 * });
 */


function Widget() {
  this.initialize.apply(this, arguments);
}

_.extend(Widget.prototype, Backbone.Events, {

    initialize: function(options) {
    var self = this;
    options = _.defaults(options, {});
    if (_.has(options, 'sync')) {
      Backbone.sync = options.sync;
    }
    this.view = new MenuView(options);

    this.listenTo(this.view, 'all', function() {
      this.trigger.apply(this, arguments);
    });
  },

});


module.exports = {
  headerObj: Widget,
  header: fs.readFileSync(__dirname + '/header.html', 'UTF8'),
  footer: fs.readFileSync(__dirname + '/footer.html', 'UTF8')
};
