// amp-settings/src/models/global-settings.js
var _ = require('underscore');
var Backbone = require('backbone');
module.exports = Backbone.Model.extend({
	url: '/rest/amp/settings',
	
	parse: function(settings){
		this.extractNumberFormatSettings(settings);
		return settings;
	},
	extractNumberFormatSettings: function(settings) {
		  var numberFormat = {}; 
	      numberFormat.numberFormat = settings['number-format'] || '#,#.#';

		  // If the format pattern doesnt have thousands grouping then ignore 'number-group-separator' param or it will 
		  // be used by JS to group by thousands (ie: in the 'Others' columns).
		  if(numberFormat.numberFormat.indexOf(',') !== -1) {			  		  
			  numberFormat.groupSeparator = settings['number-group-separator'] || ',';
		  } else {
			  numberFormat.groupSeparator = '';
		  }
		  			  
		  numberFormat.decimalSeparator = settings['number-decimal-separator'] || '.';
		  this.numberFormatSettings = numberFormat;		
		  
		  this.numberDivider = settings['number-divider'];		  
		  if (this.numberDivider === 1) {
			  this.numberDividerDescription = 'amp.dashboard:chart-tops-inunits';
		  } else if(this.numberDivider === 1000) {
			  this.numberDividerDescription = 'amp.dashboard:chart-tops-inthousands';
		  } else if(this.numberDivider === 1000000) {
			  this.numberDividerDescription = 'amp.dashboard:chart-tops-inmillions';
		  }else if(this.numberDivider === 1000000000) {
			  this.numberDividerDescription = 'amp.dashboard:chart-tops-inbillions';
		  }
	  }

});
