// amp-filter/src/tree/tree-node-model.js
var _ = require('underscore');
var Backbone = require('backbone');
var TreeNodeModel; // declare here to help with ref loop of collection and model
var TreeNodeCollection = Backbone.Collection.extend({model: TreeNodeModel});
var Constants = require('../utils/constants');

var hasAnySelectedChildren = false;

TreeNodeModel = Backbone.Model.extend({
    defaults: {
        selected: undefined,  // default is selected. change to string / ternary, for off, semi, and on
        expanded: true,
        visible: true,
        numSelected: 0,
        numPossible: 0,
        children: null,     // type TreeNodeCollection
        isSelectable: false,  // is this node itself selectable (ie. should it have an 'unkown' child),
        reset: null,
        parent: undefined,
        'include-location-children': undefined
    },

    // For deserialization purposes.
    ignoreOrgGroupList: ['implementing-agency',
        'beneficiary-agency',
        'executing-agency',
        'responsible-organization',
        'component-funding-organization',
        'component-second-responsible-organization'],

    initialize: function (obj) {
        var self = this;
        var childrenCollection = new TreeNodeCollection();
        this.set('children', childrenCollection);
        if (Array.isArray(obj.children)) {
            _.each(obj.children, function (child) {
                child.parent = self;
                var newChild = new TreeNodeModel(child);
                childrenCollection.add(newChild);
            });
        }
        if (!childrenCollection.isEmpty() && this.get('isSelectable')) {
            var node = new TreeNodeModel(self.toJSON());
            node.set('name', this.get('name'));
            childrenCollection.add(node);
        }
        this._addListenersToChildren();
        this.on('change:selected', self._onSelectChange);
        this._updateCount();
    },


    /* This function works from top to bottom of the tree. */
    serialize: function (options) {
        // This variable is a constant now just in case in the far future we need some/all filters to send not only
        // the parent but its children too.
        var stopOnParent = true;
        var isLocationAndDontIncludeChildren = this._isLocationAndDontIncludeChildren();
        // Location is a special case because we could select a parent without its children.
        if (isLocationAndDontIncludeChildren === true) {
            stopOnParent = false;
        }
        var tmpSerialized = {};
        var children = this.get('children');
        // AMP-28683: Before checking if it has children, check if the "parent" is selected.
        // IMPORTANT: we cant trust on "selected" field on branch nodes, except in the special case for locations
        // without including children. The problem is branch nodes only have selected===true when you click on them.
        if (options.includeUnselected ||
            (isLocationAndDontIncludeChildren === false && this.get('numSelected') === this.get('numPossible')) ||
            (isLocationAndDontIncludeChildren === true && this.get('selected'))) {
            tmpSerialized[this.get('filterId')] = (options.wholeModel ? [this] : [this.id]);
            if (options.includeUnselected || (!stopOnParent && children.length > 0)) {
                this._serializeChildren(tmpSerialized, children, options);
            }
        } else {
            // If current node is not selected and it has children.
            if (children.length > 0) {
                // AMP-28671: if not including location with children then serialize everything selected.
                if (isLocationAndDontIncludeChildren === true) {
                    this._serializeChildren(tmpSerialized, children, options);
                } else if (this.get('numSelected') === this.get('numPossible')) {
                    tmpSerialized[this.get('filterId')] = (options.wholeModel ? [this] : [this.id]);
                } else {
                    this._serializeChildren(tmpSerialized, children, options);
                }
            }
        }
        return tmpSerialized;
    },

    _serializeChildren: function (tmpSerialized, children, options) {
        if (children.length > 0) {
            children.each(function (child) {
                var serializedChild = child.serialize(options);
                _.each(serializedChild, function (v, k) {
                    if (v) {
                        //merge if key exists
                        if (tmpSerialized[k]) {
                            tmpSerialized[k] = tmpSerialized[k].concat(v);
                        } else {
                            tmpSerialized[k] = v;
                        }
                    }
                });
            });
        }
    },

    deserialize: function (blob) {
        var self = this;
        var children = this.get('children');
        if (children.length > 0) {
            children.each(function (child) {
                if (!child.get('filterId') && self.get('filterId')) {
                    child.set('filterId', self.get('filterId'));
                }
                child.deserialize(blob);
            });
        }
        if (blob[this.get('filterId')]) {
            if (this.isInFilters(blob, this.id)) {
                this.set('selected', true, {propagation: true});
            } else if (children.length === 0) {
                this.set('selected', false, {propagation: true});
            }
        }
    },

    isInFilters: function (blob, id) {
        var result = false;
        if (!(this.ignoreOrgGroupList.indexOf(this.get('filterId')) > -1 && this.get('isOrgGroup') === true)) {
            result = (blob[this.get('filterId')]).indexOf(this.id) > -1;
        }
        return result;
    },

    _onSelectChange: function (model, argument, options) {
        var self = this;
        var children = this.get('children');
        if (this.get('selected')) {
            this.set('numSelected', this.get('numPossible'));
        } else {
            if (this._isLocationAndDontIncludeChildren() === true && this._hasAnySelectedChildren()) {
                this.set('numSelected', 1);
            } else {
                this.set('numSelected', 0);
            }
        }
        if (!children.isEmpty() && this._isLocationAndDontIncludeChildren() === false) {
            self._updateChildNodes(options.propagation);
        }
        if (options.propagation) {
            self.trigger('updateCount');
        }
    },

    _isLocation: function () {
        if (this.get('originalName') === Constants.LOCATIONS || this.get('originalName') === Constants.PLEDGES_LOCATIONS ||
            this.get('filterId') === Constants.COMMUNAL ||
            this.get('filterId') === Constants.DISTRICT || this.get('filterId') === Constants.PLEDGES_DISTRICT ||
            this.get('filterId') === Constants.ZONE || this.get('filterId') === Constants.PLEDGES_ZONE ||
            this.get('filterId') === Constants.REGION || this.get('filterId') === Constants.PLEDGES_REGION ||
            this.get('filterId') === Constants.COUNTRY || this.get('filterId') === Constants.PLEDGES_COUNTRY) {
            return true;
        } else {
            return false;
        }
    },

    _isLocationAndDontIncludeChildren: function () {
        // By default is undefined because we need to get the deserealized data to know if it will be true|false.
        var result = undefined;
        if (this._isLocation()) {
            // Check if the field exists but only for root level.
            if (this.get('id') === -1 && this.get('parent') === undefined) {
                if (this.get('include-location-children') !== undefined) {
                    result = !this.get('include-location-children');
                }
            } else {
                var node = this._getRootNode();
                if (node.get('include-location-children') !== undefined) {
                    result = !node.get('include-location-children');
                }
            }
        } else {
            result = false;
        }
        return result;
    },

    _getRootNode: function () {
        var isRoot = (this.get('id') === -1 && this.get('parent') === undefined);
        var node = null;
        var currentNode = this;
        while (isRoot === false) {
            if (currentNode) {
                if (currentNode.get('id') === -1 && currentNode.get('parent') === undefined) {
                    node = currentNode;
                    isRoot = true;
                } else {
                    currentNode = currentNode.get('parent');
                }
            } else {
                /* When deserealizing saved filters that happens before the tree is created so parent field is undefined
                at this point. */
                break;
            }
        }
        return node;
    },

    _hasAnySelectedChildren: function () {
        hasAnySelectedChildren = false;
        this._findChildrenSelected(this);
        return hasAnySelectedChildren;
    },

    _findChildrenSelected: function (object) {
        var self = this;
        if (hasAnySelectedChildren) {
            return true;
        }
        // Check for branch nodes first (this).
        if (object.get('selected') === true) {
            hasAnySelectedChildren = true;
            return hasAnySelectedChildren;
        }
        if (object.get('children').length === 0) {
            if (object.get('selected') === true) {
                hasAnySelectedChildren = true;
            }
            return hasAnySelectedChildren;
        } else {
            _.find(object.get('children').models, function (item) {
                return self._findChildrenSelected(item);
            });
        }
    },

    _addListenersToChildren: function () {
        var self = this;
        var children = this.get('children');
        children.each(function (child) {
            child.on('change:visible', function () {
                // If no children are visible, then hide self.
                if (!children.findWhere({visible: true})) {
                    self.set({visible: false});
                } else {
                    self.set({visible: true});
                }
            });
            child.on('updateCount', function () {
                self._updateCount();
                self.trigger('updateCount');
            });
        });
    },

    _updateCount: function () {
        var children = this.get('children');
        var countTotal = {
            selected: 0,
            possible: 0
        };

        if (!children.isEmpty()) {
            children.each(function (child) {
                countTotal.selected += child.get('numSelected');
                countTotal.possible += child.get('numPossible');
            });
            if (this._isLocationAndDontIncludeChildren() === true && this.get('selected') && countTotal.selected === 0) {
                countTotal.selected = 1;
            }
        } else {
            countTotal = {
                selected: (this.get('selected') ? 1 : 0),
                possible: 1
            };
        }

        this.set('numSelected', countTotal.selected);
        this.set('numPossible', countTotal.possible);
    },

    _updateChildNodes: function (propagation, selected) {
        var self = this;
        var children = this.get('children');

        if (!children.isEmpty()) {
            children.each(function (child) {
                child.set('selected', (selected ? selected : self.get('selected')), {propagation: propagation});
            });
        }
    },

    filterText: function (txt) {
        var children = this.get('children');

        if (!children.isEmpty()) {
            // if the node itself matches, turn it on, even if no children are visible.
            if (this.get('name') && this.get('name').toLowerCase().indexOf(txt) > -1) {
                this.set('visible', true);
            }

            // FYI: putting inside an else above keeps children visible if parent matches.
            children.each(function (child) {
                child.filterText(txt);
            });

        } else {
            if (this.get('name') && this.get('name').toLowerCase().indexOf(txt) > -1) {
                this.set('visible', true);
            } else {
                this.set('visible', false);
            }
        }
    }

});

module.exports = TreeNodeModel;
