Index: sites/all/modules/contrib/ctools/includes/context-task-handler.inc
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- sites/all/modules/contrib/ctools/includes/context-task-handler.inc	(date 1411417777000)
+++ sites/all/modules/contrib/ctools/includes/context-task-handler.inc	(revision )
@@ -178,7 +178,9 @@
     if (!empty($links) && is_array($links)) {
       $build = array(
         '#theme_wrappers' => array('container'),
-        '#attributes' => array('class' => array('contextual-links-region')),
+        '#attributes' => array('class' => array(
+          // 'contextual-links-region'
+        )),
       );
 
       if (!is_array($info['content'])) {
@@ -188,16 +190,16 @@
         $build['content'] = $info['content'];
       }
 
-      $build['contextual_links'] = array(
-        '#prefix' => '<div class="contextual-links-wrapper">',
-        '#suffix' => '</div>',
-        '#theme' => 'links__contextual',
-        '#links' => $links,
-        '#attributes' => array('class' => array('contextual-links')),
-        '#attached' => array(
-          'library' => array(array('contextual', 'contextual-links')),
-        ),
-      );
+      // $build['contextual_links'] = array(
+      //   '#prefix' => '<div class="contextual-links-wrapper">',
+      //   '#suffix' => '</div>',
+      //   '#theme' => 'links__contextual',
+      //   '#links' => $links,
+      //   '#attributes' => array('class' => array('contextual-links')),
+      //   '#attached' => array(
+      //     'library' => array(array('contextual', 'contextual-links')),
+      //   ),
+      // );
       $info['content'] = $build;
     }
   }
