Index: sites/all/modules/contrib/entityform/entityform.module
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- sites/all/modules/contrib/entityform/entityform.module	(revision 30644)
+++ sites/all/modules/contrib/entityform/entityform.module	(revision )
@@ -1210,6 +1210,11 @@
         return html_entity_decode(entity_i18n_string($name, $property_value, $langcode), ENT_QUOTES);
       }
     }
+    // @HACK: Translate entityform labels.
+    else if ($property_info['label'] == 'Label') {
+      return t($property_value);
+    }
+
     return $property_value;
   }
 
