// amp-filter/src/models/years-filter-model.js
var $ = require('jquery');
var BaseFilterModel = require('../models/base-filter-model');


module.exports = BaseFilterModel.extend({

  //TODO: serialize-deserialize functions

  defaults: {
    selectedStart: null,
    selectedEnd: null,
    // range is provided by api, but will fallback to this if ot provided, or set to -1
    startYear: '01/01/1961',
    endYear: '31/12/2015'
  },

  initialize: function(options) {
    BaseFilterModel.prototype.initialize.apply(this, [options]);
    this.url = options.endpoint;
    this.set('_loaded', $.Deferred());
  },

  parse: function(data) {
    if (!data.startYear || data.startYear === -1) {
      data.startYear = this.defaults.startYear;
    } else {
      data.startYear = '01/01/' + data.startYear;
    }
    if (!data.endYear || data.endYear === -1) {
      data.endYear = this.defaults.endYear;
    } else {
      data.endYear = '31/12/' + data.endYear;
    }

    this.get('_loaded').resolve();
    return data;
  },

  serialize: function() {
    if(this.get('selectedStart')){
    	  var key = this.get('column')!='N/A' ? this.get('column'):'date';
    	  var obj = {};
    	  obj[key] = {  
				  start: this._dateConvert(this.get('selectedStart')),
				  end: this._dateConvert(this.get('selectedEnd'))
				  };
    	  return obj;
    } else {
      return null;
    }
  },

  deserialize: function(obj) {
	if(obj && (obj[this.get('column')] || obj.date)){
	  var key = 'date'; 
	  if(obj[this.get('column')]) {
		  key = this.get('column');
	  }
	  this.set('selectedStart', this._dateConvert(obj[key].start));
      this.set('selectedEnd', this._dateConvert(obj[key].end));
    }
  },

  reset: function() {
    this.set('selectedStart', this.get('startYear'));
    this.set('selectedEnd', this.get('endYear'));
  },

  // converts: 03/01/1961 <===> 1961-01-01
  // amp expects with '-' and jQuery wants with '/'
  _dateConvert: function(input){
    var output = null;
    if(input){
      if(input.indexOf('/')>-1){
        input = input.split('/');
        output = input[2] + '-' + input[1] + '-' + input[0];
      } else if(input.indexOf('-')>-1){
        input = input.split('-');
        output = input[2] + '/' + input[1] + '/' + input[0];
      }
    }
    return output;
  }

});
