// amp-filter/src/views/filters-view.js
var fs = require('fs');
var $ = require('jquery');
var _ = require('underscore');
var Backbone = require('backbone');

var TopLevelFilterView = require('../views/top-level-filter-view');
var AllFilterCollection = require('../collections/all-filters-collection');


var Template = fs.readFileSync(__dirname + '/../templates/filters-content-template.html', 'utf8');
var TitleTemplate = fs.readFileSync(__dirname + '/../templates/filter-title-template.html', 'utf8');

module.exports = Backbone.View.extend({
  id: 'tool-filters',
  name: 'Filters',
  apiURL: '/rest/filters',

  events: {
    'click .apply': 'applyFilters',
    'click .cancel': 'cancel',
    'click .reset': 'resetFilters'
  },

  // collection of top-level-filter views..
  filterViewsInstances:{},

  template: _.template(Template),
  titleTemplate: _.template(TitleTemplate),

  initialize:function(options) {
    this.draggable = options.draggable;
    this.translator = options.translator;

    this.firstRender = true;

    this.allFilters = new AllFilterCollection();
    this._loaded =  this.allFilters._loaded;

    // Create top level views
    this._createTopLevelFilterViews();
  },

  _createTopLevelFilterViews: function() {
    this.filterViewsInstances = {
      donors: new TopLevelFilterView({name:'Donor'}),
      sectors: new TopLevelFilterView({name:'Sector'}),
      programs: new TopLevelFilterView({name:'Programs'}),
      activity: new TopLevelFilterView({name:'Activity'}),
      allAgencies: new TopLevelFilterView({name:'AllAgencies'}),
      financials: new TopLevelFilterView({name:'Financial'}),
      others: new TopLevelFilterView({name:'Other'})
    };
  },


  render: function() {
    var self = this;
    this.$el.addClass('panel panel-primary');
    if (this.draggable) {
      this.$el.draggable({ cancel: '.panel-body, .panel-footer', cursor: 'move'  });
    }

    if (this.firstRender) {
      this.$el.html(this.template({}));
      this.$el.show();

      this._getFilterList().done(function() {
        self.renderFilters();

        // setup any popovers as needed...
        self.popovers = self.$('[data-toggle="popover"]');
        self.popovers.popover();

        // Translate if available.
        if (self.translator) {
          self.translator.translateDOM(self.el);
        }
      });

      this.firstRender = false;
    }

    return this;
  },


  renderFilters: function() {
    this.$('.filter-options').html('');

    for (var filterView in this.filterViewsInstances) {
      if (this.filterViewsInstances.hasOwnProperty(filterView)) {
        var tmpFilterView = this.filterViewsInstances[filterView];
        this.$('.filter-titles').append(tmpFilterView.renderTitle().titleEl);

        //...render bodies on click, not all at once...doesn't seem critical right now...
        this.$('.filter-options').append(tmpFilterView.renderFilters().el);
      }
    }

    this.$('.filter-titles a:first').tab('show');
  },


  _getFilterList: function() {
    var self = this;
    return this.allFilters.load().then(function() {
      self.allFilters.each(function(model) {
        self._createFilterViews(model);
      });
      return this;
    });
  },

  _createFilterViews: function(tmpModel) {

    // TODO: magic strings are dangerous, config somewhere...
    switch (tmpModel.get('group')) {
      case 'ActivityBudgetList':
      case 'TypeOfAssistanceList':
      case 'FinancingInstrumentsList':
        this.filterViewsInstances.financials.filterCollection.add(tmpModel);
        break;
      case 'ActivityStatusList':
      case 'ActivityApprovalStatus':
        this.filterViewsInstances.activity.filterCollection.add(tmpModel);
        break;
      case 'Programs':
        this.filterViewsInstances.programs.filterCollection.add(tmpModel);
        break;
      case 'Sectors':
        this.filterViewsInstances.sectors.filterCollection.add(tmpModel);
        break;
      case 'Donor':
        this.filterViewsInstances.donors.filterCollection.add(tmpModel);
        break;
      case 'Role':
        this.filterViewsInstances.allAgencies.filterCollection.add(tmpModel);
        break;
      default:
        this.filterViewsInstances.others.filterCollection.add(tmpModel);
    }

  },


  serialize: function() {
    var self = this;
    this.serializedFilters = {};

    this.allFilters.each(function(filter) {
    //TODO: build a util for bettermerge that concat's array if duplicate keys in objects...
      _.extend(self.serializedFilters, filter.serialize());
    });

    //remove empty / false values.
    _.each(this.serializedFilters, function(v, k) {
      if(!v || _.isEmpty(v)) {
        delete self.serializedFilters[k];
      }
    });

    return this.serializedFilters;
  },

  deserialize: function(blob, options) {
    if (blob) {
      this.allFilters.each(function(filter) {
        filter.deserialize(blob);
      });
    } else {
      console.warn('could not deserialize blob:', blob);
    }
    if (!options || !options.silent) {
      this.applyFilters();  // triggers the "apply" event
    }
  },

  showFilters: function() {
    this.render();
    this.filterStash = null;  // in case they haven't loaded yet, don't try to .serialize()
    this._loaded.done(_.bind(function() { this.filterStash = this.serialize(); }, this));
  },

  resetFilters: function() {
    this.allFilters.each(function(filter) {
      filter.reset();
    });
  },

  applyFilters: function() {
    var state = this.serialize();
    this.trigger('apply', state);
  },

  cancel: function() {
    this.deserialize(this.filterStash, {silent: true});
    this.trigger('cancel', this.filterStash);
  }
});
