// gis-layers-manager/src/views/admin-view.js
var fs = require('fs');
var $ = require('jquery');
var _ = require('underscore');
var Backbone = require('backbone');
var LayerIndicatorCollection = require('../collections/layers-collections');
var Events = require('../utils/events.js');

var Template = fs.readFileSync(__dirname + '/../templates/admin-template.html', 'utf8');
module.exports = Backbone.View.extend({  
  id: 'layer-manager-admin',
  events: {
      'click .glyphicon-edit': 'editLayer',
      'click .glyphicon-remove': 'removeLayer',
      'click .page-item': 'changePage',
      'click .new-layer-btn': 'createLayer'
  },
  template: _.template(Template),  
  initialize:function(options) {
      this.layers = new LayerIndicatorCollection();
      this.EventsBus = options.EventsBus;
      this.translate = options.translate;
      this.translator = options.translator;
      var self = this;
      this.layers.fetch({          
          success: function (collection, response, options) {
              self.render();
          },
          error: function(){
        	  console.log('error loading indicators');
          }
      });
      /*this.layers.add([
          {id: 1, name: "Indicator 1", username: "atl", creationDate: "06/10/2016", type: "Private","unit":"%", admLevelId: 77, accessTypeId: 'Private', colorRamp: 2, description : "description description description 1"},
          {id: 2, name: "Indicator 2", username: "guess", creationDate: "08/10/2016", type: "Public", "unit":"%", admLevelId: 78, accessTypeId: 'Private', colorRamp: 3, description : "description description description 2"}
      ]);*/
      this.page = new Backbone.Model({
          pageArea: null,
          recordsPerPage: 10,
          currentPageNumber: 2,
          totalPageCount: 3,
          totalRecords: 25
      })
  },
  render: function() {
      this.$el.html(this.template({layers: this.layers, page: this.page}));
      this.translate(this.$el);
      this.$el.show();
      return this;
  },
  editLayer: function(e) {
      var id = $(e.target).data("id");
      var selectedModel = this.layers.findWhere({id: id});      
      this.EventsBus.trigger(Events.UPDATE_LAYER_EVENT, selectedModel);
  },
  createLayer: function() {           
	  this.EventsBus.trigger(Events.CREATE_LAYER_EVENT);
  },
  removeLayer: function(e) {
      if(confirm("Are you sure you want to delete")) {
          var id = $(e.target).data("id");
          var selectedModel = this.layers.findWhere({id: id});
          selectedModel.delete();
      }
  },
  changePage: function(e) {
      debugger;
      var page = $(e.target).data("page");
      if(page === '+') {
          page = this.layers.page + 1;
      } else if(page === '-') {
          page = this.layers.page - 1;
      }
      this.layers.page = page;
      this.layers.fetch();
      this.render();
  }
});
