// gis-layers-manager/src/views/main-view.js
var fs = require('fs');
var $ = require('jquery');
var _ = require('underscore');
var Backbone = require('backbone');
var Translator = require('amp-translate');
var Wizard = require('./wizard');
var Admin = require('./admin');

var Template = fs.readFileSync(__dirname + '/../templates/main-view.html', 'utf8');
module.exports = Backbone.View.extend({
  id: 'layer-manager',
  events: {  
	  'click .cancel': 'cancel'
  },
  template: _.template(Template),  
  initialize:function(options) {
	  this.draggable = options.draggable;
	  this.caller = options.caller;
	  this.wizard = new Wizard({});
	  this.admin = new Admin({});
  },  
  render: function() {
	    var self = this;
	    this.$el.addClass('panel panel-primary');
	    if (this.draggable) {
	      this.$el.draggable({ cancel: '.panel-body, .panel-footer', cursor: 'move', containment: 'window' });
	    }
	    this.$el.html(this.template({}));
	    this.$el.show();
	    this.$el.find("#wizard").empty();
	    this.$el.find("#admin").empty();
	    this.$el.find("#wizard").append(this.wizard.render().el);
	    this.$el.find("#admin").append(this.admin.render().el);
	    return this;
	},	  
	show: function(){
		this.render();
	},
	cancel: function(){
		this.trigger('cancel');
	}
});
