// gis-layers-manager/src/views/main-view.js
var fs = require('fs');
var $ = require('jquery');
var _ = require('underscore');
var Backbone = require('backbone');
var Translator = require('amp-translate');
var Wizard = require('./wizard-view');
var Admin = require('./admin-view');
var PopulationLayerView = require('./population-layer-view');
var Translator = require('amp-translate');
var IndicatorLayer = require('../models/indicator-layer.js');
var Template = fs.readFileSync(__dirname + '/../templates/main-template.html', 'utf8');
var EventsBus = _({}).extend(Backbone.Events);
var Events = require('../utils/events.js');
var SettingsCollection = require('../collections/settings-collection');
module.exports = Backbone.View.extend({
	id: 'layer-manager',
	events: {  
		'click .cancel': 'cancel'
	},
	template: _.template(Template),  
	initialize:function(options) {
		
		_.bindAll(this, 'render');
		var self = this;
		this.draggable = options.draggable;
		this.caller = options.caller;
		this.settings = new SettingsCollection();
		if(options.translator === undefined) {
			this.createTranslator(true);
		} else {
			this.translator = options.translator;
		}		
		
		EventsBus.on(Events.UPDATE_LAYER_EVENT, function(selected) {
			self.wizard = new Wizard({translate: self.translate,translator: self.translator, model: selected, EventsBus: EventsBus, settings: self.settings});
			self.$el.find("#layers-wizard").html(self.wizard.render().el);
			self.$el.find("#layers-admin").hide();
			self.$el.find("#layers-wizard").show();
		});
		
		EventsBus.on(Events.CREATE_LAYER_EVENT, function() {
			var model = new IndicatorLayer({name:{}});
			self.wizard = new Wizard({translate: self.translate,translator: self.translator, model: model, EventsBus: EventsBus, settings: self.settings});
			self.$el.find("#layers-wizard").html(self.wizard.render().el);
			self.$el.find("#layers-admin").hide();
			self.$el.find("#layers-wizard").show();
		});
		
		EventsBus.on(Events.WIZARD_CLOSED_EVENT,function(){
			self.$el.find("#layers-wizard").hide();
			self.$el.find("#layers-admin").empty();		
			self.admin = new Admin({translate: self.translate,translator: self.translator, EventsBus: EventsBus, settings: self.settings});
			self.$el.find("#layers-admin").html(self.admin.render().el);
			self.admin.delegateEvents();
			self.$el.find("#layers-admin").show();				
		});	
		
		EventsBus.on(Events.OPEN_POPULATION_VIEW_EVENT, function() {			
			self.$el.find("#layers-wizard").hide();
			self.$el.find("#layers-admin").hide();		
			self.populationLayerView = new PopulationLayerView({translate: self.translate,translator: self.translator, EventsBus: EventsBus, settings: self.settings});
			self.$el.find("#population-layer-view").html(self.populationLayerView.render().el);
			self.populationLayerView.delegateEvents();
			self.$el.find("#population-layer-view").show();			
		});
		
		EventsBus.on(Events.POPULATION_VIEW_CLOSED_EVENT, function() {			
			self.$el.find("#population-layer-view").hide();
			self.$el.find("#layers-admin").show();	
		});		
		 
	},  
	render: function() {
		var self = this;
		this.deffered = [];
	    this.deffered.push(this.settings.fetch());
	    $.when.apply($, this.deffered).then(function () {
	    	self.$el.addClass('panel panel-primary');
			if (self.draggable) {
				self.$el.draggable({cursor: 'move', containment: 'window' });
			}
			self.$el.html(self.template({}));
			self.$el.show();	
			self.$el.find("#layers-admin").empty();
			self.admin = new Admin({translate: self.translate,translator: self.translator, EventsBus: EventsBus,settings: self.settings});
			self.$el.find("#layers-admin").append(self.admin.render().el);
			self.admin.delegateEvents();
			self.translate(self.$el);
	    });		
		return this;
	},
	createTranslator: function(force) {
		var self = this;
		var translateKeys = JSON.parse(fs.readFileSync(__dirname + '/../lib/initial-translation-request.json', 'utf8'));
		// setup any popovers as needed...
		//self.popovers = self.$('[data-toggle="popover"]');
		//self.popovers.popover();
		if (force === true || self.translator === undefined) {	      
			self.translator = new Translator({defaultKeys: translateKeys});
		}
	},
	translate: function(target) {
		var element = this;
		if (target !== undefined) {
			element = target;
		}
		if (element.el !== undefined) {
			this.translator.translateDOM(element.el);
		} else {
			this.translator.translateDOM(element);
		}
	},
	show: function(){
		this.render();
	},
	cancel: function(){
		this.trigger('cancel');
	}
});
