// gis-layers-manager/src/views/wizard-view.js
var fs = require('fs');
var $ = require('jquery');
var _ = require('underscore');
var Backbone = require('backbone');
var ColorCollection = require('../collections/colors-collection.js');
var ADMCollection = require('../collections/adm-levels-collection.js');
var AccessTypesCollection = require('../collections/access-types-collection.js');
var Steps = require('../utils/steps.js');
var Validator = require('../utils/validator.js');
var Events = require('../utils/events.js');
var Template = fs.readFileSync(__dirname + '/../templates/wizard-template.html', 'utf8');
var IndicatorLayer = require('../models/indicator-layer.js');
module.exports = Backbone.View.extend({
  id: 'layer-manager-wizard',  
  events: {	
	  'click .next-button ': 'next',
	  'click .previous-button': 'previous',	  
	  'click .btn-back': 'backToList',
	  'change .form-control':  'onInputChange',
	  'click .color-radio': 'selectColors',
	  'click .btn-submit' : 'submitLayer'	  
  },  
  template: _.template(Template),  
  initialize:function(options) {	
	  _.bindAll(this, 'render', 'next','previous','onInputChange','loadCollections','submitLayer','backToList');
	  this.EventsBus = options.EventsBus;
	  this.translate = options.translate;
	  this.translator = options.translator;
	  var self = this;	   
	  this.loadCollections().then(function() {	
		  if(!_.isUndefined(options.model)){
			  self.model = options.model;			  
			  self.render();
		  }		  
	  });
	 
  },  
  render: function() {	 
	    var self = this;
	    this.$el.html(this.template(this.model.toJSON()));
	    this.renderColorSelector();
	    this.renderADMDropdown();
	    this.renderAccessTypeDropdown();
	    this.translate(this.$el);
	    this.$el.show();
	    return this;
  },  
  renderColorSelector: function(){
	  var self = this;
	  this.colors.each(function(colorModel){	    	
	    	_.each(colorModel.attributes,function(colorArray, key){
	    		 var tr = $('<tr></tr>');
	    		 var colortd = $('<td></td>');
	    		 var radiotd = $('<td></td>');
	    		 var colorStrip = $('<div class="dd-option-description dd-desc"></div>');
	    		_.each(colorArray,function(hexColor){
	    			var colorSpan = $('<span></span>');
	    			colorSpan.html('&nbsp;&nbsp;&nbsp;');
	    			colorSpan.css({'background-color':hexColor});
	    			colorStrip.append(colorSpan);    			
	    		});
	    		var radio = $('<input class="color-radio"/>');
	    		radio.attr('type','radio')
	    		radio.attr('value',key);
	    		radio.attr('name','color-input');
	    		if(key == self.model.get('colorRamp')){
	    			radio.prop('checked', true);
	    		}    		
	    		tr.append(radiotd.append(radio));
	    		tr.append(colortd.append(colorStrip));
	    		self.$el.find(".color-selector").append(tr);
	    	});
	    });  
  },
  renderADMDropdown: function(){
	  var self = this;
	  this.admLevels.each(function(admLevel){
		   var selected = (admLevel.get('id') === self.model.get('admLevelId'));	   
	       self.$el.find(".adm-level").append($("<option></option>").attr("value",admLevel.get('id')).prop('selected', selected).text(admLevel.get('label')));	    	
	    });  
  },
  renderAccessTypeDropdown: function(){
	  var self = this;	  
	  this.accessTypes.each(function(accesType){
		    var selected = (accesType.get('accessType') === self.model.get('accessTypeId'));		    
	    	self.$el.find(".access-type").append($("<option></option>").attr("value",accesType.get('accessType')).prop('selected', selected) .text(accesType.get('accessType')));	    	
	    });
  },
  next: function(e){     
	  var target = e.target;
	  var current = $(target).attr("data-current");		  
	  var errors = Validator[current](this.model);  
	  if(errors.length === 0){
		  this.$el.find('.nav-pills a[href="#' + Steps[current].next + '"]').tab('show'); 
	  }else{
		  this.displayErrors(errors);
	  }           	  
  },
  //TODO translation
  displayErrors: function(errors){	  
	  var messageBox = this.$el.find('.message-box');
	  messageBox.empty();	  
	_.each(errors, function(error){
		var span = $('<span></span>');		
		span.html(error);
		messageBox.append($('<span class="glyphicon glyphicon-exclamation-sign"></span>'));
		messageBox.append(span)
		messageBox.append($('<br>'));
	});  
	messageBox.show().delay(5000).fadeOut();
  },
  previous: function(e){
	  var target = e.target;
	  var current = $(target).attr("data-current");	  	 
	  if(Steps[current].previous != current){		    
		  this.$el.find('.nav-pills a[href="#' + Steps[current].previous + '"]').tab('show');
	  }	 
  },  
  onInputChange: function(e){
	  var fieldName = $(e.target).attr('id')	  
	  this.model.set(fieldName, $(e.target).val());	  
  },
  
  loadCollections: function(){
	  var deffereds = [];
	  this.colors = new ColorCollection();
	  this.admLevels = new ADMCollection();
	  this.accessTypes = new AccessTypesCollection();    
	  deffereds.push(this.colors.fetch());
	  deffereds.push(this.admLevels.fetch());
	  deffereds.push(this.accessTypes.fetch());
	  return $.when.apply($, deffereds);	  
  },  
  selectColors : function(e){	  
	  this.model.set('colorRamp', $(e.target).val());	
  },
  submitLayer: function(){
	  var self = this;
	  this.model.save(null, {
		    success: function (model, response) {
		    	self.EventsBus.trigger(Events.WIZARD_CLOSED_EVENT);
		    },
		    error: function (model, response) {
		    	self.EventsBus.trigger(Events.WIZARD_CLOSED_EVENT);		        
		    }
		});	  
  },  
  backToList : function(){
	  this.EventsBus.trigger(Events.WIZARD_CLOSED_EVENT);
  }
});
