// gis-layers-manager/src/models/indicator-layer.js
var _ = require('underscore');
var Backbone = require('backbone');
module.exports = Backbone.Model.extend({
  urlRoot: '/rest/indicator/indicator-layer/',
  defaults: {
	    name: {},
	    unit: '',
	    description: {},
	    numberOfClasses: 0
  },
  initialize: function(){
  },
  url: function(){
	  return '/rest/indicator/indicator-layer/';
  },
  parse : function(resp, xhr) {
	 var data =(resp.data) ? resp.data: resp;
	 data.name = this.parseJSON(data.name);
	 data.description = this.parseJSON(data.description);
     return data;
  },
  parseJSON: function(str){
	 try {
	        return JSON.parse(str);
	    } catch (e) {
	        return str;
	    }
   }, 
   deleteLayer: function(options){
	 //TODO change to use backbone destroy. destroy fails because of incorrect url
	 $.ajax({
         url : this.url() + this.id,
         type : 'DELETE',
         success :options.success
  });
 },
 update: function(success) {
     return this.save(null, {
         type: 'POST',
         success: success
     });

 }
});
