// amp-boilerplate/src/views/header-view.js
var fs = require('fs');
var Backbone = require('backbone');
var _ = require('underscore');
require('bootstrap/dist/js/bootstrap');
var Template = fs.readFileSync(__dirname + '/../templates/header-template.html', 'utf8');

module.exports = Backbone.View.extend({
  model: null,
  template: _.template(Template),
  el: '#amp-footer',  
  showAdminFooter: true,
  showDGFooter: true,  
  events: {
	    "click .user-url": "showUserProfile"
  },
  initialize: function(options) {
    this.showAdminFooter = options.showAdminFooter;
    this.showDGFooter = options.showDGFooter;
    //this.model = options.model;
    var self = this;
    $.when(options.layoutFetched).then(function() {
    	self.render();
    });  
    _.bindAll(this, 'render', 'refreshUserSection');
  },
  render: function() {
    if (this.model) {
      this.refreshUserSection();
      var self = this;
      this.$el.html(this.template({
        properties: self.model,
        showAdminLinks: self.showAdminFooter,
        showDGFooter: self.showDGFooter
      }));
    }
    return this;
  },
  refreshUserSection: function() {
	var self = this;
    if (this.model.logged === true) {
      $('.container-fluid', $('#amp-header')).toggleClass('ampUserLoggedIn');
      $('#headerUrl').attr("href", "/aim");
    } else {
      $('#headerUrl').attr("href", "/");
    }
    if (this.model.email) {    	 
      $("#header-name" ).click(function() {		  
		  self.showUserProfile();
      });
      $('#header-workspace', $('#amp-header')).text(this.model.workspace);
      $('#header-workspace', $('#amp-header')).prop('title', this.model.workspace);
      $('#header-name #header-first-name', $('#amp-header')).text(this.model.firstName);
      $('#header-name #header-last-name', $('#amp-header')).text(this.model.lastName);
    }

  },  
  showUserProfile: function(){
	if (this.model.email) {
       var url = '/aim/default/userProfile.do~edit=true~id=' + this.model.userId;       
       if(!this.profileLoaded ){    	   
          $('.user-profile-content').load(url);  
          this.profileLoaded = true;
       }            
       if (typeof $().modal !== 'function') {
    	      $.noConflict();
       }       
       $('#user-profile').modal({
    	      show: true,
    	      backdrop: false
      });
	}
  }

});
