// amp-boilerplate/src/views/header-footer-view.js
var fs = require('fs');
var $ = require('jquery');
var Backbone = require('backbone');
var _ = require('underscore');

var Template = fs.readFileSync(__dirname + '/../templates/footer-template.html', 'utf8');
var LayoutModel = require('../models/amp-layout-model.js');

module.exports = Backbone.View.extend({
	  model : null,
	  template: _.template(Template),
	  el: '#amp-footer',
	  layoutFetched : new $.Deferred(),
	  showAdminFooter:true,

	  initialize: function(options) {
		this.showAdminFooter = options.showAdminFooter;
	    var layoutModel = new LayoutModel();
	    var self = this;
	    layoutModel.fetch().then(function(layout) {
	    	self.model = layout;
	    	window.buildDate = layout.buildDate;
	    	window.ampVersion = layout.ampVersion;
	    	self.render();
	    	self.layoutFetched.resolve();
	    });
	    _.bindAll(this, 'render','refreshUserSection');
	 
	  },
	  render: function() {
		  if (this.model) {
			  this.refreshUserSection ();
			  var self = this;
			  this.$el.html(this.template({
		       properties: self.model,
		       showAdminLinks:self.showAdminFooter
		      }));
		  }
	      return this;
	    },
	  refreshUserSection:function () {
		  if (this.model.logged === true) {
			    $('.container-fluid',$('#amp-header')).toggleClass('ampUserLoggedIn');
		  }
		  if (this.model.email) {
		        $('#header-workspace', $('#amp-header')).text(this.model.workspace);
		        $('#header-name #header-first-name', $('#amp-header')).text(this.model.firstName);
		        $('#header-name #header-last-name', $('#amp-header')).text(this.model.lastName);
		  }
	    	
	    }

	});
