// amp-boilerplate/src/views/header-footer-view.js
var fs = require('fs');
var Backbone = require('backbone');
var _ = require('underscore');

var Template = fs.readFileSync(__dirname + '/../templates/footer-template.html', 'utf8');

var UserModel = require('../models/amp-user-model.js');

module.exports = Backbone.View.extend({
	  model : null,
	  template: _.template(Template),
	  appendEl: '#amp-footer',

	  initialize: function(options) {
	    var userModel = new UserModel();
	    var $footer = $('#amp-footer');
	    window.boilerf = $footer;

	    //$header.html(headerWidget.view.render().el);
	    var self = this;
	    userModel.fetch().then(function(user) {
	    	self.model = user;
	    	self.render();
	    });
	    _.bindAll(this, 'render');
	 
	  },
	  render: function() {
		  if (this.model.email) {
		        $('.container-fluid',$('#amp-header')).toggleClass('ampUserLoggedIn');
		        $('#header-workspace', $('#amp-header')).text(this.model.workspace);
		        $('#header-name #header-first-name', $('#amp-header')).text(this.model.firstName);
		        $('#header-name #header-last-name', $('#amp-header')).text(this.model.lastName);
		  }
		  var self = this;
		  this.$el.html(this.template({
	       properties: self.model,
	      }));
	      $(this.appendEl).append(this.$el.html());
	      return this;
	    }


	});
