// amp-boilerplate/src/views/header-footer-view.js
var fs = require('fs');
var Backbone = require('backbone');
var _ = require('underscore');
require('bootstrap/dist/js/bootstrap');

var Template = fs.readFileSync(__dirname + '/../templates/footer-template.html', 'utf8');
var LayoutModel = require('../models/amp-layout-model.js');

module.exports = Backbone.View.extend({
  model: null,
  template: _.template(Template),
  el: '#amp-footer',
  layoutFetched: new $.Deferred(),
  showAdminFooter: true,
  showDGFooter: true,  
  events: {
	    "click .user-url": "showUserProfile"
  },
  initialize: function(options) {
    this.showAdminFooter = options.showAdminFooter;
    this.showDGFooter = options.showDGFooter;
    var layoutModel = new LayoutModel();
    var self = this;
    layoutModel.fetch().then(function(layout) {
      self.model = layout;
      window.buildDate = layout.buildDate;
      window.ampVersion = layout.ampVersion;
      self.render();
      self.layoutFetched.resolve();
    });

    //AMP-20646: we need to wait until the endpoint has responded.
    this.layoutFetched.done(function() {
      self.render();
    });
    _.bindAll(this, 'render', 'refreshUserSection');
  },
  render: function() {
    if (this.model) {
      this.refreshUserSection();
      var self = this;
      this.$el.html(this.template({
        properties: self.model,
        showAdminLinks: self.showAdminFooter,
        showDGFooter: self.showDGFooter
      }));
    }
    return this;
  },
  refreshUserSection: function() {
	var self = this;
    if (this.model.logged === true) {
      $('.container-fluid', $('#amp-header')).toggleClass('ampUserLoggedIn');
    }
    if (this.model.email) {    	 
      $("#header-name" ).click(function() {		  
		  self.showUserProfile();
      });
      $('#header-workspace', $('#amp-header')).text(this.model.workspace);
      $('#header-workspace', $('#amp-header')).prop('title', this.model.workspace);
      $('#header-name #header-first-name', $('#amp-header')).text(this.model.firstName);
      $('#header-name #header-last-name', $('#amp-header')).text(this.model.lastName);
    }

  },  
  showUserProfile: function(){
	if (this.model.email) {
       var url = '/aim/default/userProfile.do~edit=true~id=' + this.model.userId;       
       if(!this.profileLoaded ){    	   
          $('.user-profile-content').load(url);  
          this.profileLoaded = true;
       }            
       if (typeof $().modal !== 'function') {
    	      $.noConflict();
       }       
       $('#user-profile').modal({
    	      show: true,
    	      backdrop: false
      });
	}
  }

});
