// amp-boilerplate/src/views/menu-view.js
var fs = require('fs');
var Backbone = require('backbone');
var _ = require('underscore');

var Template = fs.readFileSync(__dirname + '/../templates/menu-template.html', 'utf8');

var UserModel = require('../models/amp-user-model.js');
var MenuCollection = require('../collections/amp-menus-collection.js');
var MenuModel = require('../models/amp-menus-model.js');

var SubmenuView = require('./submenu-compositeview.js');

module.exports = Backbone.View.extend({
  id: 'menus',

  events: {
    'click .apply': 'applyFilters',
    'click .reset': 'resetFilters'
  },
  appendEl: '#AmpMenus',

  template: _.template(Template),

  initialize: function(options) {
    this.collection = new MenuCollection();
    window.boilers = this;
    this.collection.fetch().then(function() {window.boilers.addAll()});

    this.firstRender = true;
    _.bindAll(this, 'addOne', 'addAll');

  },
  addAll: function() {
    this.collection.each(this.addOne);
  },

  addOne: function(model) {
    view = new SubmenuView({model: model});
    view.render();
    $(this.appendEl).append(view.el);
  },

  render: function() {
      var self = this;
      this.$el.addClass('');

      if (this.firstRender) {
        this.$el.html(this.template({}));

        this.addAll();
        this.$el.show();
        $('.dropdown-toggle').dropdown()

        this.firstRender = false;
      }

      return this;
    }


});
