// amp-boilerplate/src/views/menu-view.js
var fs = require('fs');
var $ = require('jquery');
var Backbone = require('backbone');
require('bootstrap/dist/js/bootstrap');
var _ = require('underscore');

var Template = fs.readFileSync(__dirname + '/../templates/menu-template.html', 'utf8');

var UserModel = require('../models/amp-user-model.js');
var MenuCollection = require('../collections/amp-menus-collection.js');
var MenuModel = require('../models/amp-menus-model.js');

var SubmenuView = require('./submenu-compositeview.js');

module.exports = Backbone.View.extend({
  el: '#amp-header',
  events: {
    'click .apply': 'applyFilters',
    'click .reset': 'resetFilters'
  },
  appendEl: '#AmpMenus',
  template: _.template(Template),
  menuRendered : new $.Deferred(),

  
  initialize: function(options) {
    this.collection = new MenuCollection();
    var self = this;
    this.collection.fetch().then(function() {self.render();});
    this.firstRender = true;
    _.bindAll(this, 'addOne', 'addAll');

  },
  addAll: function() {
    this.collection.each(this.addOne);
    this.menuRendered.resolve();
  },

  addOne: function(model) {
    view = new SubmenuView({model: model});
    view.render();
    $(this.appendEl).append(view.el);
  },

  render: function() {
      if (this.firstRender) {
        this.$el.html(this.template({}));
        this.addAll();
        //$('.dropdown-toggle').dropdown();
        this.firstRender = false;
      }

      return this;
    }


});
