// amp-boilerplate/src/views/menu-view.js
var fs = require('fs');
var $ = require('jquery');
var Backbone = require('backbone');
require('bootstrap/dist/js/bootstrap');
var _ = require('underscore');

var Template = fs.readFileSync(__dirname + '/../templates/menu-template.html', 'utf8');
var MenuCollection = require('../collections/amp-menus-collection.js');
var MenuModel = require('../models/amp-menus-model.js');

var SubmenuView = require('./submenu-compositeview.js');
var AboutView = require('./about-view.js');


module.exports = Backbone.View.extend({
  el: '#amp-header',
  appendEl: '#AmpMenus',
  template: _.template(Template),
  menuRendered : new $.Deferred(),
 
  initialize: function(options) {
    this.collection = new MenuCollection();
    var self = this;
    this.collection.fetch().then(function() {self.render();});
    this.firstRender = true;
    this.about = new AboutView(options);
    _.bindAll(this, 'addOne', 'addAll','showAbout');

  },
  
  addAll: function() {
    this.collection.each(this.addOne);
    this.menuRendered.resolve();
  },

  addOne: function(model) {
    view = new SubmenuView({model: model});
    this.listenTo(view, 'showAbout', this.showAbout);
    var self = this;
    this.listenTo(view, 'switchLanguage', function(lng) {
    	self.trigger ('switchLanguage',lng);
    });
    view.render();
    $(this.appendEl).append(view.el);
  },

  render: function() {
      if (this.firstRender) {
        this.$el.html(this.template({}));
        this.addAll();
        this.firstRender = false;
      }

      return this;
  },
  showAbout: function() {
     if ($('#about-popup').length == 0) {
		  this.$el.parent().append(this.about.render().el);
	   }
	  $('#about-popup').modal({show: true, backdrop: false});
	  return false;
  }

});
